### PR TITLE
Specify Custom End Dates + Union Bugfixes

### PR DESCRIPTION
**Pull request by chakrabortyr**
This is #602 over again, wanted to make sure everything worked correctly so I didn't spam you guys with commits.

## Description
When ingesting a discrete set of cloud data, we may run into the issue of having a start date for a VM but no clearly defined end. To remediate this issue, we enable specification of a custom default end date as a string in the etl process.

To use, specify the -d option with etl overseer. Some example commands:

`etl_overseer -r RESOURCE -e 'now - 7 days' -p cloud-state-pipeline`
`etl_overseer -r RESOURCE -e '2018-02-26 + 1 second' -p cloud-state-pipeline`
`etl_overseer -r RESOURCE -e '+ 14 days' -p cloud-state-pipeline`

## Motivation and Context
See above.

## Tests performed
Manual verification of data consistency, added component test to validate event reconstruction.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by smgallo** [comment]
The ETL overseer already has an `-e | --end-date` option why not use that?

File: open_xdmod/modules/xdmod/component_tests/lib/ETL/CloudStateReconstructorTransformIngestorTest.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    );
+
+    private $fsm;
+
+    public function __construct()
+    {
+        $configFile = realpath(BASE_DIR . '/tests/artifacts/xdmod-test-artifacts/xdmod/etlv2/configuration/input/xdmod_etl_config_8.0.0.json');
+
+        // This needs to be explicitly defined here so PHP 5.4 doesn't complain
+        $definitionLoc = BASE_DIR . "/tests/artifacts/xdmod-test-artifacts/xdmod/etlv2/configuration/input/etl_action_defs_8.0.0.d/cloud_state.json";
```
Comment (smgallo) [inline open_xdmod/modules/xdmod/component_tests/lib/ETL/CloudStateReconstructorTransformIngestorTest.php]: Rather than create a single use variable combine these 2 lines:
```
$this->options_array["definition_file"] =
    BASE_DIR .
    "/tests/artifacts/xdmod-test-artifacts/xdmod/etlv2/configuration/input/etl_action_defs_8.0.0.d/cloud_state.json";
```

File: classes/CCR/Loggable.php [hunk 2]
```diff
@@ -129,7 +129,10 @@ public function logAndThrowException($message, array $options = null)
 
         $logMessage['message'] = $message;
 
-        $this->logger->log($logMessage, $logLevel);
+        if ($this->logger) {
```
Comment (smgallo) [inline classes/CCR/Loggable.php]: Why is this needed? If a logger has not been passed to the constructor it is set to `Logger::singleton('null')` in `setLogger()`.

Comment (chakrabortyr) [reply classes/CCR/Loggable.php]: I added this after manually running tests and seeing the Logger barf when the null Logger was used specifically in that function. Adding the conditional seemed to fix the issue, so I left it there as an added safety check.

Comment (smgallo) [reply classes/CCR/Loggable.php]: I have not run into the issue. Have you looked at my working tests? I'd like to remove the conditional unless it is needed.

File: classes/ETL/Ingestor/CloudStateReconstructorTransformIngestor.php [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    /**
+     * @see ETL\Ingestor\pdoIngestor::transform()
+     */
+    protected function transform(array $srcRecord, &$orderId)
+    {
+        // We want to just flush when we hit the dummy row
+        if ($srcRecord['event_type_id'] === 0) {
+            return array($this->_instance_state);
```
Comment (jpwhite4) [inline classes/ETL/Ingestor/CloudStateReconstructorTransformIngestor.php]: need an
```php
if ($this->instance_state !== null) {
   return array($this->_instance_state);
} else {
   return array();
}
```

File: open_xdmod/modules/xdmod/component_tests/phpunit.xml.dist [hunk 4]
```diff
@@ -6,8 +6,8 @@
          cacheTokens="true"
          colors="true"
          convertErrorsToExceptions="true"
-         convertNoticesToExceptions="true"
-         convertWarningsToExceptions="true"
+         convertNoticesToExceptions="false"
```
Comment (jpwhite4) [inline open_xdmod/modules/xdmod/component_tests/phpunit.xml.dist]: Why did this change?

File: classes/CCR/Loggable.php [hunk 5]
```diff
@@ -130,6 +130,7 @@ public function logAndThrowException($message, array $options = null)
         $logMessage['message'] = $message;
 
         $this->logger->log($logMessage, $logLevel);
+
```
Comment (jpwhite4) [inline classes/CCR/Loggable.php]: don't make changes that just add whitespace to files that would otherwise be not modified.

File: classes/ETL/Ingestor/CloudStateReconstructorTransformIngestor.php [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+     * @see ETL\Ingestor\pdoIngestor::__construct()
+     */
+    public function __construct(aOptions $options, EtlConfiguration $etlConfig, Log $logger = null)
+    {
+        parent::__construct($options, $etlConfig, $logger);
+
+        $this->_stop_event_ids = array(self::STOP, self::TERMINATE, self::SUSPEND, self::SHELVE);
+        $this->_start_event_ids = array(self::START, self::RESUME, self::STATE_REPORT, self::UNSHELVE);
+        $this->_all_event_ids = array_merge($this->_start_event_ids, $this->_stop_event_ids);
+        $this->_end_time = $etlConfig->getVariableStore()->endDate;
```
Comment (jpwhite4) [inline classes/ETL/Ingestor/CloudStateReconstructorTransformIngestor.php]: You should do the 
```php
date('Y-m-d H:i:s', strtotime($etlConfig->getVariableStore()->endDate))
```
call here so that it only has to happen once and not every time there is a new instance.

File: open_xdmod/modules/xdmod/component_tests/phpunit.xml.dist [hunk 7]
```diff
@@ -6,8 +6,8 @@
...
+         convertWarningsToExceptions="false"
```
Comment (jpwhite4) [inline open_xdmod/modules/xdmod/component_tests/phpunit.xml.dist]: Please revert these changes.
